### PR TITLE
Adding Weather Loader

### PR DESCRIPTION
**Pull request by 0xcaffeinated**
# Adding Weather Loader

## Before submitting

Below is the recording of the module testing:
https://github.com/hwchase17/langchain/assets/69910091/2b179eec-1421-4509-9ac8-89f57f739ba6

## Who can review?
@eyurtsev

### CODE REVIEW

**Comment by eyurtsev** [comment]
@iamadhee code is looking pretty good! I added a few minor comments to avoid overloading `load`. I'd recommend providing an implementation for `lazy_load` by default rather than for `load`. if you're able it would be nice to include a test. Ideally a test that mocks the client response and does not use the network to fetch data.

**Comment by iamadhee** [comment]
> @iamadhee code is looking pretty good! I added a few minor comments to avoid overloading `load`. I'd recommend providing an implementation for `lazy_load` by default rather than for `load`. if you're able it would be nice to include a test. Ideally a test that mocks the client response and does not use the network to fetch data.

Hey @eyurtsev, I have added the `lazy_load` method, as discussed.

File: langchain/document_loaders/weather.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+            if res.forecast_minutely:
+                info_dict['minutely forecast'] = [i.to_dict() for i in res.forecast_minutely]
+            if res.national_weather_alerts:
+                info_dict['national weather alerts'] = [i.to_dict() for i in res.national_weather_alerts]
+            
+            yield Document(page_content=str(info_dict),metadata=metadata)        
+
+    def load(
+        self, 
+        places: List[str],
```
Comment (eyurtsev) [inline langchain/document_loaders/weather.py:77]:
```suggestion
```

File: langchain/document_loaders/weather.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        self.places = places
+
+    def lazy_load(
+        self,
+    ) -> Iterator[Document]:
+        """A lazy loader for document content."""
+
+        try:
+            import pyowm
+        except:
```
Comment (leo-gan) [inline langchain/document_loaders/weather.py:35]:
```except ImportError:
   raise ImportError(..
```

File: langchain/document_loaders/weather.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        mgr = owm.weather_manager()
+        reg = owm.city_id_registry()
+
+        for place in self.places:
+            info_dict = {}
+            list_of_locations = reg.locations_for(city_name=place)
+
+            try:
+                city = list_of_locations[0]
+            except:
```
Comment (leo-gan) [inline langchain/document_loaders/weather.py:48]: `except ValueError:`

**Comment by iamadhee** [comment]
@dev2049 Added the sample notebook as discussed

File: langchain/document_loaders/weather.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+    Reads the forecast & current weather of any location using OpenWeatherMap's free
+    API. Checkout 'https://openweathermap.org/appid' for more on how to generate a free
+    OpenWeatherMap API.
+    """
+
+    def __init__(
+        self,
+        client: OpenWeatherMapAPIWrapper,
+        places: Iterable[str],
```
Comment (eyurtsev) [inline langchain/document_loaders/weather.py:23]: Might make sense to use a Sequence here -- i think it's reasonable to expect that we're dealing with a finite number of places that's loaded into memory

**Review by eyurtsev — APPROVED** [review]

File: langchain/document_loaders/weather.py (hunk 2 above)
Comment (leo-gan) [reply langchain/document_loaders/weather.py]: Please, change
```
except: 
   raise ValueError(
```
to
```
except ImportError:
   raise ImportError(

```

**Review by leo-gan — COMMENTED** [review]
One nit: error handling on the `import`. Otherwise, lgtm.

**Pull request merged**